### ISSUE TITLE
Bound properties in Puppets gets called several times during a request.

### ISSUE THREAD

**Issue by chrhol** [Starcounter/Starcounter#2551]
Bound properties in Json gets called too many times during a request. Most likely cause is that the dirtycheck checks the value against the cached value and sets the cached value more than once.

**Comment by chrhol** [comment]
Will start to investigate and (hopefully) fix this issue.

**Comment by chrhol** [comment]
Other things and issues have taken focus from this one, so I have not gotten as far as I wanted to. 

The reason for the excessive accessing is the dirtycheck together with the serializer, that retrieves it several times. 

So a "normal" execution when jsonpatches are created looks something like this. 

1. Gathering changes that will be made into jsonpatches. Bound value is read and compared to stored unbound value. If not equal a change is added. This step might be skipped though, if the jsonproperty is already marked as dirty.
2. The needed size for all patches is estimated to be able to allocate one buffer that can hold everything. Here the value is read again.
3. Patches are written. Third time the value is read.
4. The jsonproperty is checkpointed, which means all flags (dirty, modified, etc) is reset, and the current value (for bound properties) are once last time read and stored in the unbound part.

Now, if the bound value is a simple getter (for example accessing a property that reads an int from the database) performance wise it shouldn't matter that it is accessed 3-4 times. The value needs to be stored and read from somewhere anyway however it is implemented. The problem comes when the bound property does other stuff, for example creates new CLR-objects that are used as views (which is not a particulary good pattern, and one we probably need to find some other solution for).

The solution I can come up with is to cache bound values when generating jsonpatches. The value for bound properties are already partly cached, since this is how the dirtycheck keeps track of changes (i.e. reading the bound value and storing it as an unbound value).

So in the first step the bound value will be read and stored and a flag will be set for that property. After that only the unbound value will be used until the all patches are created.

This will mean that bound value will be accessed once during this time, instead of three or four times. However this will only help during this sequence. Accessing the property from applicationcode or calling for example `json.ToJson` directly will still result in more then one read.

**Comment by chrhol** [comment]
The implementation for this is now pushed to develop and pnext. There was quite a lot of changes so I will continue tomorrow to do some additional testing to make sure it works as intended, and that we don't have any regressions in other functionality.

**Comment by miyconst** [comment]
Does this mean the the bound properties in JSON will be called only once? Can I test it now?

**Comment by chrhol** [comment]
> Does this mean the the bound properties in JSON will be called only once? Can I test it now?

Yes, once per request (when patches are generated). Unless you call it yourself from app.

There were no nightly that succeeded though, so you have to create an installer from the buildserver.

**Comment by miyconst** [comment]
This sounds good, but I will wait for an installer.

**Comment by chrhol** [comment]
I found one place I forgot to change, when the whole viewmodel is returned in the first request. Will fix and push this soon.

**Comment by chrhol** [comment]
Missed fix pushed now to both branches.

**Comment by chrhol** [comment]
Looks like there might still be some problems with this implementation. An issue where patches are not sent was reported from Tunity. 

I have reproduced it, and is currently trying to simplify the case when it happens.

**Comment by chrhol** [comment]
I was too fast in commenting here. The issue has nothing to do with this implementation. Moving it to a separate issue.

**Comment by chrhol** [comment]
I will consider this issue resolved for now.
